jsdom-nodejs-env/src/main/scala/org/scalajs/jsenv/jsdomnodejs/JSDOMNodeJSEnv.scala

```diff
@@ -8,138 +8,210 @@
+    val command = config.executable :: config.args
+    val externalConfig = ExternalJSRun.Config()
+      .withEnv(env)
+      .withRunConfig(runConfig)
+    ExternalJSRun.start(command, externalConfig)(JSDOMNodeJSEnv.write(files))
+  }
 
-  // TODO We might want to make this configurable - not sure why it isn't
-  override protected def wantSourceMap: Boolean = false
+  private def initFiles: List[VirtualBinaryFile] =
+    List(JSDOMNodeJSEnv.runtimeEnv, Support.fixPercentConsole)
 
-  override def jsRunner(files: Seq[VirtualJSFile]): JSRunner =
-    new DOMNodeRunner(files)
+  private def env: Map[String, String] =
+    Map("NODE_MODULE_CONTEXTS" -> "0") ++ config.env
 
-  override def asyncRunner(files: Seq[VirtualJSFile]): AsyncJSRunner =
-    new AsyncDOMNodeRunner(files)
+  private def scriptFiles(input: Input): List[VirtualBinaryFile] = input match {
+    case Input.ScriptsToLoad(scripts) => scripts
+    case _                            => throw new UnsupportedInputException(input)
+  }
 
-  override def comRunner(files: Seq[VirtualJSFile]): ComJSRunner =
-    new ComDOMNodeRunner(files)
+  private def codeWithJSDOMContext(input: Input): List[VirtualBinaryFile] = {
+    val scriptsURIs = scriptFiles(input).map {
+      case file: FileVirtualFile => file.file.toURI
+      case file                  => JSDOMNodeJSEnv.materialize(file)
+    }
+    val scriptsURIsAsJSStrings =
+      scriptsURIs.map(uri => '"' + escapeJS(uri.toASCIIString) + '"')
+    val jsDOMCode = {
+      s"""
+         |(function () {
+         |  var jsdom;
+         |  try {
+         |    jsdom = require("jsdom/lib/old-api.js"); // jsdom >= 10.x
+         |  } catch (e) {
+         |    jsdom = require("jsdom"); // jsdom <= 9.x
+         |  }
+         |
+         |  var virtualConsole = jsdom.createVirtualConsole()
+         |    .sendTo(console, { omitJsdomErrors: true });
+         |  virtualConsole.on("jsdomError", function (error) {
+         |    /* This inelegant if + console.error is the only way I found
+         |     * to make sure the stack trace of the original error is
+         |     * printed out.
+         |     */
+         |    if (error.detail && error.detail.stack)
+         |      console.error(error.detail.stack);
+         |
+         |    // Throw the error anew to make sure the whole execution fails
+         |    throw error;
+         |  });
+         |
+         |  /* Work around the fast that scalajsCom.init() should delay already
+         |   * received messages to the next tick. Here we cannot tell whether
+         |   * the receive callback is called for already received messages or
+         |   * not, so we dealy *all* messages to the next tick.
+         |   */
+         |  var scalajsCom = global.scalajsCom;
+         |  var scalajsComWrapper = scalajsCom === (void 0) ? scalajsCom : ({
+         |    init: function(recvCB) {
+         |      scalajsCom.init(function(msg) {
+         |        process.nextTick(recvCB, msg);
```

Comment:
Because in the base Node.js env, there is no return to the event loop between the startup and `start()`, so the network messages haven't had a chance to be processed by a JS event handler, so they're not yet in the message buffer in JS. They're still in a C++ buffer somewhere. In the jsdom env, there is a come back to the event loop before this code is executed.

It took me a while to figure it out ^^

Reply:
Ah, ah, I see. Fair enough. So yes, it should definitely be moved to `ComSupport`.